detection_rules/rule.py

```diff
@@ -3,294 +3,290 @@
 # 2.0; you may not use this file except in compliance with the Elastic License
 # 2.0.
 """Rule object."""
-import base64
-import copy
-import hashlib
 import json
-import os
+from dataclasses import dataclass, field
 from pathlib import Path
+from typing import Literal, Union, Optional, List
 from uuid import uuid4
 
 import eql
+from marshmallow import validates_schema
 
 import kql
-from . import ecs, beats
-from .rule_formatter import nested_normalize, toml_write
-from .schemas import CurrentSchema, TomlMetadata, downgrade
+from . import ecs, beats, utils
+from .rule_formatter import toml_write
+from .schemas import downgrade
+from .schemas import definitions
 from .utils import get_path, cached
 
 RULES_DIR = get_path("rules")
 _META_SCHEMA_REQ_DEFAULTS = {}
 
 
-class Rule(object):
-    """Rule class containing all the information about a rule."""
+@dataclass(frozen=True)
+class RuleMeta:
+    """Data stored in a rule's [metadata] section of TOML."""
+    creation_date: str
```

Comment:
how does marshmallow serialize these dates on output? `YYYY/MM/DD`?

Reply:
you can override and define the format in a base schema

Reply:
otherwise default uses `-`. But also we treat date as a string, so converting it datetime could be problematic, which is why I recommended the `definitions.Date`

Reply:
i used `definitions.Date` for now that does the regex.
we can explore a legit date later if we want
a00af312701569b6ce41ba83e1dcaf8ed318e2a4